panic with --sql_no_caching

This command works fine without `--sql_no_caching`, but with this options it panics.
```
$ docker run -i -t --rm --name=nss -p 4222:4222 --link=mysql nats-streaming:0.10.2 --cluster_id=dev --store=SQL --sql_driver=mysql --sql_source='root:@tcp(mysql)/nss' --sql_no_caching -SDV
[1] 2018/07/12 12:48:59.676698 [INF] STREAM: Starting nats-streaming-server[dev] version 0.10.2
[1] 2018/07/12 12:48:59.676741 [INF] STREAM: ServerID: X5uCI58JEwfeWQ3ZBmprQ6
[1] 2018/07/12 12:48:59.676746 [INF] STREAM: Go version: go1.10.3
[1] 2018/07/12 12:48:59.680869 [INF] Starting nats-server version 1.1.0
[1] 2018/07/12 12:48:59.680891 [INF] Git commit [not set]
[1] 2018/07/12 12:48:59.681836 [INF] Listening for client connections on 0.0.0.0:4222
[1] 2018/07/12 12:48:59.681843 [INF] Server is ready
[1] 2018/07/12 12:48:59.706006 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-send false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/12 12:48:59.706325 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-general false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/12 12:48:59.706559 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-acks false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/12 12:48:59.706751 [INF] STREAM: Recovering the state...
[1] 2018/07/12 12:48:59.709505 [INF] STREAM: Shutting down.
[1] 2018/07/12 12:48:59.709720 [DBG] STREAM: connection "_NSS-dev-send" has been closed
[1] 2018/07/12 12:48:59.709739 [INF] Server Exiting..
[1] 2018/07/12 12:48:59.709756 [DBG] STREAM: connection "_NSS-dev-general" has been closed
[1] 2018/07/12 12:48:59.709768 [DBG] STREAM: connection "_NSS-dev-acks" has been closed
[1] 2018/07/12 12:48:59.709788 [INF] STREAM: Failed to start: runtime error: invalid memory address or nil pointer dereference
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x83f844]

goroutine 1 [running]:
github.com/nats-io/nats-streaming-server/server.RunServerWithOpts.func1(0xc420230000, 0xc4201e5f50)
	/go/src/github.com/nats-io/nats-streaming-server/server/server.go:1428 +0x1c9
panic(0x9b8ac0, 0xdfa350)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/nats-io/nats-streaming-server/stores.(*SQLSubStore).recoverPendingRow(0xc420186460, 0xc42016dc00, 0xc4201864d0, 0x0, 0xc4205e0b70, 0x0, 0x0, 0x0)
	/go/src/github.com/nats-io/nats-streaming-server/stores/sqlstore.go:1955 +0x2c4
github.com/nats-io/nats-streaming-server/stores.(*SQLStore).Recover(0xc42029e000, 0x0, 0x0, 0x0)
	/go/src/github.com/nats-io/nats-streaming-server/stores/sqlstore.go:913 +0x122c
github.com/nats-io/nats-streaming-server/server.(*StanServer).start(0xc420230000, 0x0, 0x0, 0x0)
	/go/src/github.com/nats-io/nats-streaming-server/server/server.go:1602 +0x117
github.com/nats-io/nats-streaming-server/server.RunServerWithOpts(0xc420220000, 0xc420220240, 0x0, 0x0, 0x0)
	/go/src/github.com/nats-io/nats-streaming-server/server/server.go:1521 +0xca0
main.main()
	/go/src/github.com/nats-io/nats-streaming-server/nats-streaming-server.go:153 +0x4f
```

--- a/stores/sqlstore.go
+++ b/stores/sqlstore.go
@@ -1948,12 +1948,15 @@ func (ss *SQLSubStore) recoverPendingRow(rows *sql.Rows, sub *spb.SubState, ap *
 		}
 		pendingAcks[seq] = struct{}{}
 	} else {
-		row := &sqlSubsPendingRow{
-			ID:   ss.curRow,
-			msgs: sqlSeqMapPool.Get().(map[uint64]struct{}),
+		var row *sqlSubsPendingRow
+		if ap != nil {
+			row = &sqlSubsPendingRow{
+				ID:   ss.curRow,
+				msgs: sqlSeqMapPool.Get().(map[uint64]struct{}),
+			}
+			ap.lastSent = lastSent
+			ap.prevLastSent = lastSent
 		}
-		ap.lastSent = lastSent
-		ap.prevLastSent = lastSent
 
 		if lastSent > sub.LastSent {
 			sub.LastSent = lastSent
@@ -1961,9 +1964,11 @@ func (ss *SQLSubStore) recoverPendingRow(rows *sql.Rows, sub *spb.SubState, ap *
 		if len(pendingBytes) > 0 {
 			if err := sqlDecodeSeqs(pendingBytes, func(seq uint64) {
 				pendingAcks[seq] = struct{}{}
-				row.msgsRefs++
-				row.msgs[seq] = struct{}{}
-				ap.msgToRow[seq] = row
+				if ap != nil {
+					row.msgsRefs++
+					row.msgs[seq] = struct{}{}
+					ap.msgToRow[seq] = row
+				}
 			}); err != nil {
 				return err
 			}
@@ -1972,15 +1977,17 @@ func (ss *SQLSubStore) recoverPendingRow(rows *sql.Rows, sub *spb.SubState, ap *
 			if err := sqlDecodeSeqs(acksBytes, func(seq uint64) {
 				if _, exists := pendingAcks[seq]; exists {
 					delete(pendingAcks, seq)
-					row.acksRefs++
-					ap.ackToRow[seq] = row
+					if ap != nil {
+						row.acksRefs++
+						ap.ackToRow[seq] = row
 
-					seqRow := ap.msgToRow[seq]
-					if seqRow != nil {
-						delete(ap.msgToRow, seq)
-						seqRow.msgsRefs--
-						if seqRow.msgsRefs == 0 && seqRow.acksRefs == 0 {
-							gcedRows[seqRow.ID] = struct{}{}
+						seqRow := ap.msgToRow[seq]
+						if seqRow != nil {
+							delete(ap.msgToRow, seq)
+							seqRow.msgsRefs--
+							if seqRow.msgsRefs == 0 && seqRow.acksRefs == 0 {
+								gcedRows[seqRow.ID] = struct{}{}
+							}
 						}
 					}
 				}

--- a/stores/sqlstore_test.go
+++ b/stores/sqlstore_test.go
@@ -927,6 +927,10 @@ func TestSQLSubStoreCaching(t *testing.T) {
 		t.Fatalf("Error creating store: %v", err)
 	}
 	defer s.Close()
+	info := testDefaultServerInfo
+	if err := s.Init(&info); err != nil {
+		t.Fatalf("Error on init: %v", err)
+	}
 
 	channel := "foo"
 	cs := storeCreateChannel(t, s, channel)
@@ -1004,6 +1008,46 @@ func TestSQLSubStoreCaching(t *testing.T) {
 	// Delete sub
 	storeSubDelete(t, cs, channel, subID)
 	testSQLCheckPendingRow(t, db, subID)
+
+	// Start with fresh subscription
+	subID = storeSub(t, cs, channel)
+	// Store some pending messages
+	storeSubPending(t, cs, channel, subID, 1, 2, 3)
+	storeSubAck(t, cs, channel, subID, 2)
+	storeSubFlush(t, cs, channel)
+
+	// Recover with NoCaching enabled
+	s.Close()
+	s, err = NewSQLStore(testLogger, testSQLDriver, testSQLSource, nil, SQLNoCaching(true), SQLMaxOpenConns(5))
+	if err != nil {
+		t.Fatalf("Error opening store: %v", err)
+	}
+	defer s.Close()
+	state, err := s.Recover()
+	if err != nil {
+		t.Fatalf("Error on recovery: %v", err)
+	}
+	rc := state.Channels[channel]
+	if rc == nil {
+		t.Fatalf("Expected channel %q to be recovered", channel)
+	}
+	if count := len(rc.Subscriptions); count != 1 {
+		t.Fatalf("Expected 1 subscription to be recovered, got %v", count)
+	}
+	rs := rc.Subscriptions[0]
+	if count := len(rs.Pending); count != 2 {
+		t.Fatalf("Expected 2 pending messages, got %v", count)
+	}
+	if _, ok := rs.Pending[1]; !ok {
+		t.Fatalf("Expected seq 1 to be in pending")
+	}
+	if _, ok := rs.Pending[3]; !ok {
+		t.Fatalf("Expected seq 1 to be in pending")
+	}
+	last := rs.Sub.LastSent
+	if last != 3 {
+		t.Fatalf("Expected last sent to be 3, got %v", last)
+	}
 }
 
 func TestSQLSubStoreCachingFlushInterval(t *testing.T) {

